store/store_download.go

```diff
@@ -196,7 +196,7 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 		return err
 	}
 
-	if err := s.cacher.Get(downloadInfo.Sha3_384, targetPath); err == nil {
+	if err := s.cacher.Get(downloadInfo.Sha3_384, targetPath); err == nil || errors.Is(err, os.ErrExist) {
```

Comment:
I'm slightly nervous about this, I would feel better if we would do something like:
```go
diff --git a/store/store_download.go b/store/store_download.go
index 812b8988c6..80673bc860 100644
--- a/store/store_download.go
+++ b/store/store_download.go
@@ -187,6 +187,21 @@ type DownloadOptions struct {
        LeavePartialOnError bool
 }
 
+func validateHash(w *os.File, sha3 string) error {
+       h := crypto.SHA3_384.New()
+       if _, err := w.Seek(0, os.SEEK_SET); err != nil {
+               return err
+       }
+       if _, err := io.Copy(h, w); err != nil {
+               return err
+       }
+       actualSha3 := fmt.Sprintf("%x", h.Sum(nil))
+       if sha3 != actualSha3 {
+               return HashError{w.Name(), actualSha3, sha3}
+       }
+       return nil
+}
+
 // Download downloads the snap addressed by download info and returns its
 // filename.
 // The file is saved in temporary storage, and should be removed
@@ -196,6 +211,16 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
                return err
        }
 
+       if f, err := os.Open(targetPath); err == nil {
+               if err := validateHash(f, downloadInfo.Sha3_384); err == nil {
+                       logger.Debugf("Found already valid %q", targetPath)
+                       return nil
+               }
+               if err := os.Remove(targetPath); err != nil {
+                       return err
+               }
+       }
+
        if err := s.cacher.Get(downloadInfo.Sha3_384, targetPath); err == nil {
                logger.Debugf("Cache hit for SHA3_384 …%.5s.", downloadInfo.Sha3_384)
                return nil
@@ -249,17 +274,7 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
                }
        } else {
                // we're done! check the hash though
-               h := crypto.SHA3_384.New()
-               if _, err := w.Seek(0, os.SEEK_SET); err != nil {
-                       return err
-               }
-               if _, err := io.Copy(h, w); err != nil {
-                       return err
-               }
-               actualSha3 := fmt.Sprintf("%x", h.Sum(nil))
-               if downloadInfo.Sha3_384 != actualSha3 {
-                       err = HashError{name, actualSha3, downloadInfo.Sha3_384}
-               }
+               err = validateHash(w, downloadInfo.Sha3_384)
        }
        // If hashsum is incorrect retry once
        if _, ok := err.(HashError); ok {
```
hope this gives the idea, i.e. it feels slightly dangerous to just assume if targetFile exists that it's also right (but maybe I'm missing something)?

Reply:
I considered the possibility that the file was somehow corrupted but when I checked it seemed very unlikely. AFAICT we always download files into a temporary location, check its digest and only then move it to the final location with `os.Rename`.

Reply:
maybe I'm missing something but fwiw this change feels strange, I would expect the fix to be inside the Cache code, why decide what to do with the error one level up?

Reply:
A cache hit/miss was already determined by the download code based on the error returned by the cache, so this just changes the condition were it was. That also matches the current cache interface `Get(... string) error` which exposes the error. If we want to move the fix into the cache code, then it makes sense to put the whole decision at that level and change the interface to just yes/no `Get(... string) bool`

Reply:
@pedronis do you have an opinion whether we should validate the digests when there is already a file at `targetPath`?

Reply:
`targetPath` is the final path to the mount file, it would look like `/var/lib/snapd/snaps/foo_11.snap`. Usually we download files to a temporary location and then move them to that final path so I thought we could trust any files there to be correct. Although, I think there's one case where, if the move failed, we try to copy so it might not be impossible to get a bad file, just very unlikely. I'm happy to check the digest of the existing file (like Michael wrote) if we think it's necessary. In that case, would it also make sense to check the digest of cached files?

Reply:
I'm probably missing something but wouldn't also the copy use some kind of atomic file for the destination? so that in the end there's an atomic rename done there as well? basically I feel we might want to make sure the copy works as expected instead of adding extra checks here. Basically I think for now we should assume the target is alright, but check if the copy code might need fixes

Reply:
Thanks. We usually use rename but, in the code path where we apply deltas, a normal copy is used as a fallback if the rename fails. I can open a separate PR to remove that fallback. I'll mark this as resolved then

Reply:
to be clear  my request was to double check that the copy is robust, I didn't mention to simply drop it? to drop we need to understand why is there and whether we can?